### PR TITLE
[Feature][Attention][Kernel] Add TriangleMix sparse prefill on Ascend

### LINKED ISSUE #13246: [Feature]: Add TriangleMix static sparse prefill attention for Ascend NPU
#### 背景与定位

vLLM-Ascend 已通过 [#12833](https://github.com/vllm-project/vllm-ascend/pull/12833) 集成面向 MiniMax M3 的稀疏注意力能力，包括 index scoring、top-k 选块以及 `SparseAttentionScore` 等动态块稀疏路径。因此，本 Issue 不再泛指“昇腾缺少稀疏注意力算子”，而是希望补充一种具有不同调度方式和适用场景的 **TriangleMix 静态稀疏 prefill 注意力算子**。

TriangleMix 论文：[TriangleMix: Accelerating Prefilling via Decoding-time Contribution Sparsity](https://arxiv.org/abs/2507.21526)。

#### 希望支持的功能

TriangleMix 在 prefill 阶段采用固定的 `sink + local-window + dense-tail` 注意力模式，将长序列中间区域的注意力计算从 dense 二次复杂度降低为近似线性复杂度。为了在 Ascend NPU 上兑现理论收益，不能只在 Python 层增加 mask，而需要在底层融合算子中真正跳过被稀疏化的 Q-K 分块。

期望实现包括：

- 使用 AscendC/CANN 实现 TriangleMix 静态稀疏注意力；
- 直接读取 paged KV，避免 gather、重排和重复数据搬运；
- 在单次 kernel launch 内完成静态分块调度、QK、online softmax 和 PV；
- 缓存请求 metadata，并记录每请求的 hit/fallback 原因；
- 仅在有收益的长序列 prefill 请求上启用，短序列和不支持的请求自动回退官方 FIA；
- decode 始终沿用 vLLM-Ascend 官方注意力路径。

#### 与现有稀疏注意力路径的区别

- 现有 MiniMax M3 路径属于动态块稀疏，需要 index scoring、top-k 和动态选块 metadata，并覆盖特定模型的 prefill/decode。
- TriangleMix 使用预先确定的 sink、局部窗口和 dense tail，不需要运行时 top-k 选块，主要优化兼容 dense attention 几何的长序列 prefill。
- 两者面向不同模型和稀疏策略，TriangleMix 不是对现有 `SparseAttentionScore` 的重复实现或替代。

#### 当前实现与验证

主线集成 PR：[vllm-project/vllm-ascend#13331](https://github.com/vllm-project/vllm-ascend/pull/13331)。

测试环境：Ascend 910B3、CANN 9.0.1、PyTorch 2.10.0、torch-npu 2.10.0.post2、vLLM 0.26.0。

当前结果：

- 固定长序列算子测试中，自定义稀疏注意力相对 dense attention 的算子耗时最高下降约 55%；
- 隔离进程 A-B-B-A 配对测试中，端到端 TTFT 可信降低 **9.001%**；该数字只代表当前固定模型、长度、batch 和层配置，不代表所有场景；
- NPU 正确性矩阵 20/20 通过；
- 8192-token dense 对比：max absolute error `3.2582e-4`，mean absolute error `4.1136e-5`，relative L2 error `2.3324e-3`；
- Qwen3-8B 2048-token 端到端验证中，36 个 prefill 层命中 TriangleMix，decode 回退官方路径，测试 continuation 的 dense/sparse token IDs 一致。

该路径默认关闭，通过 `additional_config.trianglemix` 显式启用，并支持层选择、crossover 阈值、严格模式及运行时统计。

### Alternatives

可以继续使用官方 FIA dense attention，或者对适配 MiniMax M3 的模型使用现有动态稀疏注意力路径。但 Python mask 无法减少底层计算量，现有动态 top-k 路径也不能直接表达 TriangleMix 的静态 sink/local-window/dense-tail 调度。

### Additional context

目标是将 TriangleMix 作为 vLLM-Ascend 内置、默认关闭的实验性 prefill 路径合入主线，而不是要求用户额外安装或维护独立插件。合并后由 vLLM-Ascend wheel 一并构建 CANN 算子，并在不支持或无收益的场景安全回退官方 FIA。

### PR DESCRIPTION
**Pull request by ackerm0n**
### What this PR does / why we need it?

This PR adds an opt-in TriangleMix sparse-prefill path for Ascend NPU. TriangleMix uses a static sink + local-window + dense-tail attention pattern, which is different from the dynamic top-k block-sparse path used by `SparseAttentionScore`.

The implementation:

- reads paged KV directly without gather/reorder copies;
- performs sparse scheduling, QK, online softmax, and PV in one AscendC launch;
- caches request metadata and records hit/fallback counters;
- routes short or unsupported requests to the existing FIA implementation;
- keeps decode on the existing official attention path.

It is enabled through `additional_config.trianglemix`; the default behavior is unchanged.

TriangleMix paper: https://arxiv.org/abs/2507.21526

Fixes #13246

### Does this PR introduce _any_ user-facing change?

Yes. It adds the optional `additional_config.trianglemix` configuration block. TriangleMix is disabled by default, so existing users see no behavior change.

Example:

```json
{
  "trianglemix": {
    "enabled": true,
    "layers": "5,7,10,15-35",
    "strict": false,
    "stats_log_interval": 100
  }
}
```

### How was this patch tested?

Environment: Ascend 910B3, CANN 9.0.1, PyTorch 2.10.0, torch-npu 2.10.0.post2, vLLM 0.26.0.

- Built the TriangleMix CANN package and the latest vLLM-Ascend Torch binding successfully.
- Unit tests: `tests/ut/attention/a2/test_trianglemix.py` — 13/13 passed.
- NPU operator correctness matrix — 20/20 passed.
- 8192-token comparison against dense attention: max absolute error `3.2582e-4`, mean absolute error `4.1136e-5`, relative L2 error `2.3324e-3`.
- Qwen3-8B end-to-end test with a 2048-token prompt: all 36 prefill layers used TriangleMix; decode used the existing official path; dense and sparse runs produced identical token IDs for the tested continuation.
- Re-applied the patch cleanly to upstream `main` at `8f3fd59a203c3b35f29b6a77f459c9a78da7d5c0`; `git diff --check` and Python syntax checks passed.

- vLLM version: v0.26.0
- vLLM main: https://github.com/vllm-project/vllm/commit/d02df748bf9efd99022f1a062597dc3cb3808485

### CODE REVIEW

**Comment by ackerm0n** [comment]
Hi @wangxiyuan @MengqingCao, could you please review this PR when you have a chance? It adds a TriangleMix static sparse-prefill path for Ascend NPU, and all current CI checks have passed. Your feedback would be greatly appreciated. Thank you!
